design/default-env.md

```diff
@@ -0,0 +1,61 @@
+6. The default mode for `cdk synth` will be `explicit`. The implication of this (detalied below) is that if users wish to synthesize an artifact from their CDK app and deploy it at a later stage, they will have to either explicitly specify `env` for all stacks or use `cdk synth --default-env=inherit|agnostic`.
+7. Make it easier to explicitly indicate that a stack is environment-agnostic. This can already be acheieved today using `{ env: { region: Aws.REGION, account: Aws.ACCOUNT_ID } }` but can be made easier with `{ env: Stack.ANY_ENV }`.
+
+## Implications
+
+It also means that `cdk synth` will fail unless all stacks have a specific environment specified (either env-specific or env-agnostic), which may be undesirable for certain use cases. 
+
+Let's look at what people use `cdk synth` for:
+
+1. **Inspection/debugging/demo**: usually this involves emitting a template output to STDOUT and inspecting it or showing it to others.
+2. **CI/CD**: using `cdk.out` (the "cloud assembly") or the templates inside it as a self-contained artifact which can later be deployed using `cdk-deploy --app cdk.out`.
+
+For the 2nd use case (CI/CD), where `cdk synth` is used during build to produce a deployment artifact, we feel that requiring explicitness is actually a benefit. Deterministic infrastructure definitions are an important tenent of the AWS CDK, and requiring that users be explicit when they deploy apps through CI/CD is actually an added value. If users wish to synthesize environment agnostic templates, they can simply specify `env: Stack.ANY_ENV` when they define their stacks.
+
+The 1st use case, however, where `cdk synth` is used to print a template to STDOUT for inspection, is something we still wish to support without requiring users to explicitly specify `env`s. 
+
+Before we decide what to do about that, we should acknowledge that the current behavior of `cdk synth` is a bit confusing today:
+
+- If the app includes more than a single stack, `cdk synth` requires that you specify the stack name in order to print to STDOUT, otherwise it just creates `cdk.out`.
+- When `cdk synth` prints to STDOUT is prints in YAML (unless `--json` is specified), while `cdk.out` always uses JSON ([#2965](https://github.com/aws/aws-cdk/issues/2965)).
+- `cdk.out` is created no matter what, for both use cases of `synth` and even if it the app was synthesized implicitly as part of `cdk deploy`. This also somewhat causes confusion.
+
+We propose to separate the use cases into two different CLI commands: `cdk synth` and `cdk build`. The first will serve the inspection use case and the second for the CI/CD use case. This way, we can decide that `--default-env` will have a different default behavior for these two commands. For `cdk synth` (STDOUT) we will default to `agnostic` and for `cdk build` we will default to `explicit`.
```

Comment:
I hate to stray from the `synth` terminology. It corresponds to what we use in the code.

I'd rather introduce a new `cdk print` or `cdk show` for the interactive case (have `cdk synth` print a message that people should be using `cdk print` instead).

Other than that, I like the proposal.

Reply:
Yeah, maybe `cdk print` for STDOUT makes more sense and leave `cdk synth` to represent the CI/CD case.